repair/row_level.cc

```diff
@@ -1799,9 +1804,14 @@ static future<stop_iteration> repair_put_row_diff_with_rpc_stream_process_op(
         return make_ready_future<stop_iteration>(stop_iteration::no);
     } else if (row.cmd == repair_stream_cmd::end_of_current_rows) {
         rlogger.trace("Got repair_rows_on_wire from peer={}, got end_of_current_rows", from);
-        return smp::submit_to(src_cpu_id % smp::count, [from, repair_meta_id, rows = std::move(current_rows)] () mutable {
+        auto fp = make_foreign(std::make_unique<repair_rows_on_wire>(std::move(current_rows)));
+        return smp::submit_to(src_cpu_id % smp::count, [from, repair_meta_id, fp = std::move(fp)] () mutable {
             auto rm = repair_meta::get_repair_meta(from, repair_meta_id);
-            return rm->put_row_diff_handler(std::move(rows), from);
+            if (fp.get_owner_shard() == engine().cpu_id()) {
+                return rm->put_row_diff_handler(std::move(*fp), from);
```

Comment:
This doesn't fix anything. *fp will be destroyed on its owner shard, but it's empty after you moved away from it. The contents will be destroyed on the current shard, after put_row_diff_handler() returns (unless it also moves it).

Reply:
See the else branch which does not move *fp. The if branch is an optimization where the submit_to submits to the *same* shard.

Reply:
@avikivity ?

Reply:
No, they are not. See: 

```
1616     // RPC handler 
1617     future<repair_rows_on_wire> get_row_diff_handler(std::unordered_set<repair_hash> set_diff, needs_all_rows_t needs_all_rows) {
1618         return with_gate(_gate, [this, set_diff = std::move(set_diff), needs_all_rows] () mutable {
1619             return get_row_diff(std::move(set_diff), needs_all_rows).then([this] (std::list<repair_row> row_diff) {
1620                 return to_repair_rows_on_wire(std::move(row_diff));
1621             });
1622         });
1623     }
```

```
1736     // RPC handler 
1737     future<> put_row_diff_handler(repair_rows_on_wire rows, gms::inet_address from) {
1738         return with_gate(_gate, [this, rows = std::move(rows)] () mutable {
1739             return apply_rows_on_follower(std::move(rows));
1740         });     
1741     }    
```

They are changed in this series. E.g., in "repair: Futurize get_row_diff to avoid stall".

Reply:
Ping.

Reply:
Queued